### PR TITLE
Report View and Edit

### PR DESCRIPTION
**Pull request by charlie-tf**
- create view and edit wrapper components for the report form
- create a report form and annual report form
- put a couple dummy fields in the forms
- allow transitioning from view to edit, locking the report
- warn if someone else is already editing (locked) the report
- allow saving and quitting editing
- allow report submission, with a confirmation dialog
- show pages of the report if annual

## View Report
image

### Annual Report
image image

## Edit report
image

### Takeover Editing
image

### Annual Report
image
image

### Submit Report
image

## TODO:
- Add actual report fields to form
- Validate form fields (make sure required fields are filled) on submit
- Poll the BE during editing for who has the lock; nav back to view if lock is lost
- Extract Dialogs into separate components

### CODE REVIEW

File: src/components/Reports/ReportEdit.tsx [hunk 1]
```diff
@@ -1,6 +1,145 @@
...
+
+  const [confirmSubmitDialogOpen, setConfirmSubmitDialogOpen] = useState(false);
+
+  const gotoReportView = async (saveChanges: boolean) => {
+    if (saveChanges && report) {
+      await ReportService.updateReport(report);
+    }
+
+    // unlock the report
+    await ReportService.unlockReport(reportIdInt);
```
Comment (constanzauanini) [inline src/components/Reports/ReportEdit.tsx:45]: Should we check if ReportService.updateReport succedded before unlocking?

File: src/components/Reports/ReportEdit.tsx [hunk 2]
```diff
@@ -1,6 +1,145 @@
...
+  const [report, setReport] = useState<Report>();
+  useEffect(() => {
+    const getReport = async () => {
+      const result = await ReportService.getReport(reportIdInt);
+      if (result.requestSucceeded) {
+        setReport(result.report);
+      }
+    };
+
+    getReport();
```
Comment (karthikbtf) [inline src/components/Reports/ReportEdit.tsx:32]: probably check for a non-empty id before calling get report?

File: src/components/Reports/ReportEdit.tsx [hunk 3]
```diff
@@ -1,6 +1,145 @@
...
+      await ReportService.updateReport(report);
+      setShowAnnual(false);
+    }
+  };
+
+  const submitReport = async () => {
+    if (report) {
+      await ReportService.updateReport(report);
+      await ReportService.submitReport(reportIdInt);
+      await ReportService.unlockReport(reportIdInt);
```
Comment (karthikbtf) [inline src/components/Reports/ReportEdit.tsx:73]: Will be good to have one BE API that can handle all this... often times user may click a button and close the browser.. or may lose internet connection in between API requests.. etc.
Generally, workflows consisting of multiple API should be batched in the BE and not FE. 
Not a show stopper.. we have other use-cases in our FE today (storage locations, add/remove orgs, etc.). Probably should revisit those at some point.

File: src/components/Reports/ReportForm.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      <Typography key='thing1'>Thing 1 for report {`${report.year}-Q${report.quarter}`}</Typography>
+      <Typography key='thing2'>Thing 2 for report {`${report.year}-Q${report.quarter}`}</Typography>
+      <Textfield
+        key='testInput'
+        placeholder={'Test Input...'}
+        label='Details'
+        id='testInput'
+        type='text'
+        disabled={!editable}
+      />
```
Comment (karthikbtf) [inline src/components/Reports/ReportForm.tsx:34]: these are all test stuff yes?

File: src/components/Reports/ReportEdit.tsx [hunk 5]
```diff
@@ -1,6 +1,145 @@
...
+  };
+
+  const handleBack = async () => {
+    if (report) {
+      await ReportService.updateReport(report);
+      setShowAnnual(false);
+    }
+  };
+
+  const submitReport = async () => {
```
Comment (karthikbtf) [inline src/components/Reports/ReportEdit.tsx:69]: this can fail, if some other user grabbed the lock (and we haven't polled yet)

**Review by karthikbtf — COMMENTED** [review]
nice progress, some comments/thoughts

File: src/components/Reports/ReportEdit.tsx (hunk 1 above)
Comment (charlie-tf) [reply src/components/Reports/ReportEdit.tsx:45]: Yes, I will have to add checking the response to service calls in many places.

File: src/components/Reports/ReportEdit.tsx (hunk 3 above)
Comment (karthikbtf) [reply src/components/Reports/ReportEdit.tsx:73]: sounds good, i understand this is WIP

**Review by karthikbtf — APPROVED** [review]
This is good for a WIP stop.

**Review by karthikbtf — APPROVED** [review]
Thanks for the changes!

**Pull request merged**